src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionReader.cs

```diff
@@ -221,36 +216,64 @@ protected object ReflectionReadValue(XmlReaderDelegator xmlReader, XmlObjectSeri
 
             if ((primitiveContract != null && primitiveContract.UnderlyingType != Globals.TypeOfObject) || nullables != 0 || type.GetTypeInfo().IsValueType)
             {
-                context.ReadAttributes(xmlReader);
-                string objectId = context.ReadIfNullOrRef(xmlReader, Globals.TypeOfString, true);
-                if (objectId != null)
+                value = ReadItemOfPrimitiveType(xmlReader, context, type, name, ns, primitiveContract, nullables);
+            }
+            else
+            {
+                value = ReflectionInternalDeserialize(xmlReader, context, null/*collectionContract*/, type, name, ns);
+            }
+
+            return value;
+        }
+
+        private object ReadItemOfPrimitiveType(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, Type type, string name, string ns, PrimitiveDataContract primitiveContract, int nullables)
+        {
+            object value;
+            context.ReadAttributes(xmlReader);
+            string objectId = context.ReadIfNullOrRef(xmlReader, type, DataContract.IsTypeSerializable(type));
+            bool typeIsValueType = type.GetTypeInfo().IsValueType;
+            if (objectId != null)
+            {
+                if (objectId.Length == 0)
                 {
-                    if (objectId.Length == 0)
+                    objectId = context.GetObjectId();
+
+                    if (!string.IsNullOrEmpty(objectId) && typeIsValueType)
+                    {
+                        throw new SerializationException(SR.Format(SR.ValueTypeCannotHaveId, DataContract.GetClrTypeFullName(type)));
+                    }
+
+                    if (primitiveContract != null && primitiveContract.UnderlyingType != Globals.TypeOfObject)
                     {
-                        objectId = context.GetObjectId();
-                        if (primitiveContract != null && primitiveContract.UnderlyingType != Globals.TypeOfObject)
-                        {
-                            value = primitiveContract.ReadXmlValue(xmlReader, context);
-                            context.AddNewObject(value);
```

Comment:
To reviewers: removing this, `context.AddNewObject(value)`, is essentially the fix. The ReadXmlValue above already adds the value object to the context. Adding the object again caused the issue.